gio/src/subclass/file_enumerator.rs

```diff
@@ -0,0 +1,395 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use glib::{prelude::*, subclass::prelude::*, translate::*, Error};
+
+use crate::{ffi, traits::FileEnumeratorExt, Cancellable, FileEnumerator, FileInfo, IOErrorEnum};
+
+// Support custom implementation of virtual functions defined in `gio::ffi::GFileEnumeratorClass` except pairs `xxx_async/xxx_finish` for which GIO provides a default implementation.
+pub trait FileEnumeratorImpl: ObjectImpl + ObjectSubclass<Type: IsA<FileEnumerator>> {
+    fn next_file(&self, cancellable: Option<&Cancellable>) -> Result<Option<FileInfo>, Error> {
+        self.parent_next_file(cancellable)
+    }
+
+    fn close(&self, cancellable: Option<&Cancellable>) -> Result<bool, Error> {
+        self.parent_close(cancellable)
+    }
+}
+
+// Support parent implementation of virtual functions defined in `gio::ffi::GFileEnumeratorClass` except pairs `xxx_async/xxx_finish` for which GIO provides a default implementation.
+pub trait FileEnumeratorImplExt: FileEnumeratorImpl {
+    fn parent_next_file(
+        &self,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<Option<FileInfo>, Error> {
+        if self.obj().is_closed() {
+            Err(Error::new::<IOErrorEnum>(
+                IOErrorEnum::Closed,
+                "Enumerator is closed",
+            ))
+        } else {
+            unsafe {
+                let data = Self::type_data();
+                let parent_class = data.as_ref().parent_class() as *const ffi::GFileEnumeratorClass;
+
+                let f = (*parent_class)
+                    .next_file
+                    .expect("No parent class implementation for \"next_file\"");
+
+                let mut error = std::ptr::null_mut();
+                let res = f(
+                    self.obj()
+                        .unsafe_cast_ref::<FileEnumerator>()
+                        .to_glib_none()
+                        .0,
+                    cancellable.as_ref().to_glib_none().0,
+                    &mut error,
+                );
+                if error.is_null() {
+                    Ok(from_glib_full(res))
+                } else {
+                    Err(from_glib_full(error))
+                }
+            }
+        }
+    }
+
+    fn parent_close(&self, cancellable: Option<&Cancellable>) -> Result<bool, Error> {
+        unsafe {
+            let data = Self::type_data();
+            let parent_class = data.as_ref().parent_class() as *const ffi::GFileEnumeratorClass;
+
+            let f = (*parent_class)
+                .close_fn
+                .expect("No parent class implementation for \"close_fn\"");
+
+            // get the corresponding object instance without checking the reference count because the object might just be finalized.
```

Comment:
`close()` is called during `finalize()`?! That means that this can't be provided as a normal virtual method then as it's unsound to access the full object during finalization (parts might already be freed, you can easily cause memory safety issues)

Reply:
>  We can't depend on developers implementing dispose for safe usage though. Calling it in dispose seems fine but you'd have to make sure this happens automatically then.

We don't have a mechanism for injecting a call inside ObjectImpl's dispose. As currently for gtk4-rs composite templates, you have to call `dispose_template` in dispose yourself.

So I rather we don't block on this given the low usage of the API and just make sure it is clearly documented. We can revisit this once we have a clear mechanism for achieving that.

Reply:
Sounds good to me but also maybe in gio itself it should already call this in dispose instead of finalize.

Reply:
Can you also add a link to that to the docs here?